metricbeat/module/vsphere/datastorecluster/datastorecluster.go

```diff
@@ -84,18 +94,45 @@ func (m *DatastoreClusterMetricSet) Fetch(ctx context.Context, reporter mb.Repor
 	}()
 
 	var datastoreCluster []mo.StoragePod
-	err = v.Retrieve(ctx, []string{"StoragePod"}, nil, &datastoreCluster)
+	err = v.Retrieve(ctx, []string{"StoragePod"}, []string{"name", "summary", "childEntity"}, &datastoreCluster)
 	if err != nil {
 		return fmt.Errorf("error in retrieve from vsphere: %w", err)
 	}
 
+	pc := property.DefaultCollector(c)
 	for i := range datastoreCluster {
 		if ctx.Err() != nil {
 			return ctx.Err()
 		}
 
-		reporter.Event(mb.Event{MetricSetFields: m.mapEvent(datastoreCluster[i])})
+		assetNames, err := getAssetNames(ctx, pc, &datastoreCluster[i])
+		if err != nil {
+			m.Logger().Errorf("Failed to retrieve object from host %s: %w", datastoreCluster[i].Name, err)
+		}
+
+		reporter.Event(mb.Event{MetricSetFields: m.mapEvent(datastoreCluster[i], &metricData{assetNames: assetNames})})
 	}
 
 	return nil
 }
+
+func getAssetNames(ctx context.Context, pc *property.Collector, dsc *mo.StoragePod) (assetNames, error) {
+	var objects []mo.ManagedEntity
+	if len(dsc.ChildEntity) > 0 {
+		if err := pc.Retrieve(ctx, dsc.ChildEntity, []string{"name"}, &objects); err != nil {
+			return assetNames{}, err
+		}
+	}
+
+	outputDsNames := make([]string, 0, len(objects))
+	for _, ob := range objects {
+		name := strings.ReplaceAll(ob.Name, ".", "_")
+		if ob.Reference().Type == "Datastore" {
+			outputDsNames = append(outputDsNames, name)
+		}
+	}
```

Comment:
```suggestion
        outputDsNames := make([]string, 0) // Start with zero capacity
        for _, ob := range objects {
            if ob.Reference().Type == "Datastore" { // Only process Datastore objects
                name := strings.ReplaceAll(ob.Name, ".", "_")
                outputDsNames = append(outputDsNames, name)
            }
        }
```

I believe you can only append and allocate space for Datastore objects.